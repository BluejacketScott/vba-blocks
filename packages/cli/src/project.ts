import { join } from 'path';
import env from './env';
import { Config, loadConfig } from './config';
import { Manifest, Source, Reference, loadManifest } from './manifest';
import { Workspace } from './workspace';
import SourceManager, { Registration } from './sources';
import resolve, { DependencyGraph, getRegistration } from './resolve';

import { parallel } from './utils';

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
  };
}

export async function loadProject(dir?: string): Promise<Project> {
  const manifest = await loadManifest(dir || env.cwd);

  // TODO Load workspace
  const workspace: Workspace = {
    root: manifest,
    members: []
  };

  const config = await loadConfig(manifest, workspace);
  const packages = await resolve(config, workspace);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup')
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths
  };
}

export async function fetchDependencies(project: Project): Promise<Manifest[]> {
  const manager = new SourceManager(project.config);
  const manifests = await parallel(
    project.manifest.dependencies,
    async dependency => {
      const registration = getRegistration(project.packages, dependency);
      const path = await manager.fetch(registration);
      const manifest = await loadManifest(path);

      return manifest;
    }
  );

  return manifests;
}
