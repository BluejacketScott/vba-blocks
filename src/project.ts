import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
