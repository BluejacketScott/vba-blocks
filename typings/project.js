import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { loadConfig } from './config';
import { loadManifest } from './manifest';
import { loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';
import { fetchingDependencies } from './messages';
const debug = require('debug')('vba-blocks:project');
/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir = env.cwd) {
    const manifest = await loadManifest(dir);
    const config = await loadConfig();
    const workspace = await loadWorkspace(manifest);
    const lockfile = await readLockfile(workspace.root.dir);
    // Resolve packages from lockfile or from sources
    const has_dirty_lockfile = !lockfile || !(await isLockfileValid(lockfile, workspace));
    debug(!has_dirty_lockfile ? 'Loading packages from lockfile' : 'Resolving packages');
    const packages = !has_dirty_lockfile
        ? lockfile.packages
        : await resolve(config, workspace, lockfile ? lockfile.packages : []);
    const paths = {
        root: workspace.root.dir,
        dir: manifest.dir,
        build: join(manifest.dir, 'build'),
        backup: join(manifest.dir, 'build', '.backup'),
        staging: await tmpFolder({ dir: env.staging })
    };
    return {
        manifest,
        workspace,
        packages,
        config,
        paths,
        has_dirty_lockfile
    };
}
/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(project) {
    const manifests = await parallel(project.packages, async (registration) => {
        const path = await fetch(project.config.sources, registration);
        const manifest = await loadManifest(path);
        return manifest;
    }, { progress: env.reporter.progress(fetchingDependencies()) });
    return manifests;
}
/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(name, dir, options = {}) {
    dir = normalize(dir);
    const { type = 'project' } = options;
    const config = await loadConfig();
    // TODO load defaults from config
    const version = '0.0.0';
    const authors = [];
    const license = 'UNLICENSED';
    const __temp_defaults = type === 'package' ? [] : ['publish'];
    // Manually generate manifest and project
    // (may be included in project in the future)
    const manifest = {
        type,
        name,
        version,
        metadata: {
            authors,
            publish: type === 'package',
            license,
            __temp_defaults
        },
        dependencies: [],
        src: [],
        references: [],
        dir
    };
    const workspace = await loadWorkspace(manifest);
    const project = {
        manifest,
        workspace,
        packages: [],
        config,
        paths: {
            root: workspace.root.dir,
            dir,
            build: join(dir, 'build'),
            backup: join(dir, 'build', '.backup'),
            staging: await tmpFolder({ dir: env.staging })
        },
        has_dirty_lockfile: true
    };
    return project;
}
