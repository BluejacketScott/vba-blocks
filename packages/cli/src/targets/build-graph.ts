import { Source, Reference } from '../manifest';
import { Project, fetchDependencies, loadManifests } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();

  const manifests = project.manifests || (await loadManifests(project));

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
