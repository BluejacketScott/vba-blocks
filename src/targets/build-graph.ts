import { Manifest, Source, Reference } from '../manifest';
import { Project } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  dependencies: Manifest[],
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();
  const manifests = [project.manifest, ...dependencies];

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
