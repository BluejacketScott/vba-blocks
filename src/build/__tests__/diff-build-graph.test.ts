import { resolveProject } from '@vba-blocks/helpers';
import { complex as project } from '@vba-blocks/fixtures/projects';
import { loadBuildGraph } from '../build-graph';
import { diffBuildGraph } from '../diff-build-graph';
import { Component } from '../component';
import { fetchDependencies } from '../../project';

test('should diff build graph', async () => {
  const resolved = await resolveProject(project);
  const dependencies = await fetchDependencies(resolved);

  const graph = await loadBuildGraph(resolved, dependencies);

  // Add, update, and remove component
  graph.components.push(
    new Component({
      type: 'module',
      code: 'Attribute VB_Name = "added"\r\n\' added\r\n',
      source: { name: 'added', path: 'added.bas' }
    })
  );

  // TODO add src to complex project to test update and remove

  expect(diffBuildGraph(resolved, dependencies, graph)).toMatchSnapshot();
});
