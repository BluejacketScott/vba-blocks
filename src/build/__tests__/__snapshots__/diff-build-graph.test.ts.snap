// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should diff build graph 1`] = `
Object {
  "components": Object {
    "added": Array [
      Component {
        "binary": undefined,
        "code": "Attribute VB_Name = \\"added\\"
' added
",
        "metadata": Object {
          "manifest": undefined,
          "source": Object {
            "name": "added",
            "path": "added.bas",
          },
        },
        "type": "module",
      },
    ],
    "dependencies": Array [
      Component {
        "binary": undefined,
        "code": "Attribute VB_Name = \\"a\\"
' a
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [
              Object {
                "name": "d",
                "registry": "vba-blocks",
                "version": "^2.0.0",
              },
            ],
            "dir": "tests/__fixtures__/sources/a/v1-2-0",
            "name": "a",
            "package": Object {
              "authors": Array [],
              "name": "a",
              "publish": false,
              "version": "1.2.0",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "a",
                "optional": false,
                "path": "tests/__fixtures__/sources/a/a.bas",
              },
            ],
            "targets": Array [],
            "version": "1.2.0",
          },
          "source": Object {
            "name": "a",
            "optional": false,
            "path": "tests/__fixtures__/sources/a/a.bas",
          },
        },
        "type": "module",
      },
      Component {
        "binary": undefined,
        "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"b\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = True
' b
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [
              Object {
                "name": "c",
                "registry": "vba-blocks",
                "version": "^0.1.5",
              },
            ],
            "dir": "tests/__fixtures__/sources/b/v1-1-0",
            "name": "b",
            "package": Object {
              "authors": Array [],
              "name": "b",
              "publish": false,
              "version": "1.1.0",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "b",
                "optional": false,
                "path": "tests/__fixtures__/sources/b/b.cls",
              },
            ],
            "targets": Array [],
            "version": "1.1.0",
          },
          "source": Object {
            "name": "b",
            "optional": false,
            "path": "tests/__fixtures__/sources/b/b.cls",
          },
        },
        "type": "class",
      },
      Component {
        "binary": undefined,
        "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"d\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = True
' d
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [
              Object {
                "name": "f",
                "registry": "vba-blocks",
                "version": "^2",
              },
            ],
            "dir": "tests/__fixtures__/sources/d/v2-0-0",
            "name": "d",
            "package": Object {
              "authors": Array [],
              "name": "d",
              "publish": false,
              "version": "2.0.0",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "d",
                "optional": false,
                "path": "tests/__fixtures__/sources/d/d.cls",
              },
            ],
            "targets": Array [],
            "version": "2.0.0",
          },
          "source": Object {
            "name": "d",
            "optional": false,
            "path": "tests/__fixtures__/sources/d/d.cls",
          },
        },
        "type": "class",
      },
      Component {
        "binary": undefined,
        "code": "Attribute VB_Name = \\"f\\"
' f
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [
              Object {
                "name": "g",
                "registry": "vba-blocks",
                "version": "^1",
              },
            ],
            "dir": "tests/__fixtures__/sources/f/v2-0-0",
            "name": "f",
            "package": Object {
              "authors": Array [],
              "name": "f",
              "publish": false,
              "version": "2.0.0",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "f",
                "optional": false,
                "path": "tests/__fixtures__/sources/f/f.bas",
              },
            ],
            "targets": Array [],
            "version": "2.0.0",
          },
          "source": Object {
            "name": "f",
            "optional": false,
            "path": "tests/__fixtures__/sources/f/f.bas",
          },
        },
        "type": "module",
      },
      Component {
        "binary": undefined,
        "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"g\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = True
' g
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [],
            "dir": "tests/__fixtures__/sources/g/v1-0-0",
            "name": "g",
            "package": Object {
              "authors": Array [],
              "name": "g",
              "publish": false,
              "version": "1.0.0",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "g",
                "optional": false,
                "path": "tests/__fixtures__/sources/g/g.cls",
              },
            ],
            "targets": Array [],
            "version": "1.0.0",
          },
          "source": Object {
            "name": "g",
            "optional": false,
            "path": "tests/__fixtures__/sources/g/g.cls",
          },
        },
        "type": "class",
      },
      Component {
        "binary": undefined,
        "code": "Attribute VB_Name = \\"c\\"
' c
",
        "metadata": Object {
          "manifest": Object {
            "dependencies": Array [],
            "dir": "tests/__fixtures__/sources/c/v0-1-6",
            "name": "c",
            "package": Object {
              "authors": Array [],
              "name": "c",
              "publish": false,
              "version": "0.1.6",
            },
            "project": undefined,
            "references": Array [],
            "src": Array [
              Object {
                "name": "c",
                "optional": false,
                "path": "tests/__fixtures__/sources/c/c.bas",
              },
            ],
            "targets": Array [],
            "version": "0.1.6",
          },
          "source": Object {
            "name": "c",
            "optional": false,
            "path": "tests/__fixtures__/sources/c/c.bas",
          },
        },
        "type": "module",
      },
    ],
    "existing": Array [],
    "removed": Array [],
  },
  "references": Object {
    "added": Array [],
    "dependencies": Array [],
    "existing": Array [],
    "removed": Array [],
  },
}
`;
